eth/stagedsync/stage_mining_bor_heimdall.go

```diff
@@ -0,0 +1,89 @@
+package stagedsync
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ledgerwatch/log/v3"
+
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon/consensus/bor/finality/whitelist"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/erigon/dataflow"
+	"github.com/ledgerwatch/erigon/eth/stagedsync/stages"
+)
+
+func MiningBorHeimdallForward(
+	ctx context.Context,
+	cfg BorHeimdallCfg,
+	stageStage *StageState,
+	unwinder Unwinder,
+	tx kv.RwTx,
+	logger log.Logger,
+) error {
+	if cfg.borConfig == nil || cfg.heimdallClient == nil {
+		return nil
+	}
+
+	logPrefix := stageStage.LogPrefix()
+	headerStageProgress, err := stages.GetStageProgress(tx, stages.Headers)
+	if err != nil {
+		return err
+	}
+
+	header := cfg.miningState.MiningBlock.Header
+	headerNum := header.Number.Uint64()
+	if headerNum <= headerStageProgress {
+		return fmt.Errorf("attempting to mine %d, which is behind current head: %d", headerNum, headerStageProgress)
+	}
+
+	// Whitelist service is called to check if the bor chain is on the canonical chain according to milestones
+	whitelistService := whitelist.GetWhitelistingService()
+	if whitelistService != nil && !whitelistService.IsValidChain(headerNum, []*types.Header{header}) {
```

Comment:
@manav2401 blocks are broadcast to the p2p network when they are freshly out of the mining loop (in addition to being communicated to the sync loop) - think this check is here to prevent the miner from broadcasting a bad block, ive left it in to maintain same logic as before my refactor

Reply:
Alright yeah makes sense. Although, it's very rare for this to happen but if the node is out of sync and still tries to produce a block, it may happen. Thanks!

Reply:
yes, we can always revise this later once we do bigger refactors - will keep it in mind